edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnection.cs

```diff
@@ -93,15 +94,15 @@ public async Task<ICloudProxy> CreateOrUpdateAsync(IClientCredentials newCredent
                             if (newCredentials is ITokenCredentials tokenAuth && this.tokenGetter.HasValue)
                             {
                                 if (IsTokenExpired(tokenAuth.Identity.IotHubHostName, tokenAuth.Token))
-                                {                                    
+                                {
                                     throw new InvalidOperationException($"Token for client {tokenAuth.Identity.Id} is expired");
                                 }
 
                                 this.tokenGetter.ForEach(tg =>
                                  {
-                                     tg.SetResult(tokenAuth.Token);
+                                     // First reset the token getter and then set the result.
                                      this.tokenGetter = Option.None<TaskCompletionSource<string>>();
-                                     Events.NewTokenObtained(newCredentials.Identity.IotHubHostName, newCredentials.Identity.Id, tokenAuth.Token);
+                                     tg.SetResult(tokenAuth.Token);
```

Comment:
Why does this order matter?

Reply:
Because the code below expects that the tokenGetter will be reset when the value is set. And that also answers your other question, I suppose

Reply:
In principle, if order is important, then we should do these under a lock. Otherwise order shouldn't matter.

Reply:
As discussed offline, the order is important because the caller is waiting for SetResult to be called (not looping), and when it resumes, it expects this.tokenGetter to be reset.